### ISSUE TITLE
Generate occupancy grid for potholes

### ISSUE THREAD

**Issue by AbBaSaMo** [Monash-Connected-Autonomous-Vehicle/ESDA#12]
- [x] use the zed cam to detect potholes simulated as white circles on the ground
- [ ] convert the result of the previous stage to an occupancy grid and output to topic potholes/map

**Comment by AnthonyZhOon** [comment]
I think this is not directly writing a costmap 2d layer, it seems like we can make another occupancy_grid_camera topic and add it with a occupancy grid generated by lidar.  This might require writing an occupany grid publisher for both lidar and camera sensors. Then the costmap is generated the same for both lidar and camera obstacles.

**Comment by AbBaSaMo** [comment]
@AnthonyZhOon hmm, so do you mean rather than create costmap layers, we fuse our multiple occupancy grids and publish the result to the /map topic which nav2 uses?

So fuse:
- slam output
- lane detection
- pothole detection

**Comment by AnthonyZhOon** [comment]
I think it makes more sense to identify all obstacles there as the costmap is a postprocessing on these obstacles to create a heatmap. though there is an argument that only hard walls are lethal obstacles and lanes can be weaker obstacles

**Comment by AbBaSaMo** [comment]
@AnthonyZhOon makes sense. In that case I'll update the issues to reflect this as generating occ grids to be fused.

**Comment by AbBaSaMo** [comment]
Actually gonna close the issue rather than delete as this can all be reflected and discussed in this one https://github.com/Monash-Connected-Autonomous-Vehicle/ESDA/issues/25 and 2 is now redundant rather than 1

**Comment by AbBaSaMo** [comment]
Re-opening issue and re-purposing it to fusion of the 3 occupancy grids:

- [ ] have a topic potholes/map
- [ ] have a topic lanes/map
- [ ] have a topic remapped from slam_tb slam/map
- [ ] fuse the 3 into topic /map

**Comment by AnthonyZhOon** [comment]
Written a node for fusing 2 occupancy grids in the occupancy-fuser branch. Need to test it still but had trouble showing grids in rviz2.
It creates a new grid based on the largest width and height of collected grids, and uses a 2d convolve to add each subscribed grid to the centre of the new grid.

**Comment by AbBaSaMo** [comment]
@AnthonyZhOon any errors being output by rviz or nodes or screenshots? could it be a missing transform issue?

I can take a look monday after 4. Either way moving issue from in progress to review.

**Comment by AbBaSaMo** [comment]
@AnthonyZhOon are there unpushed changes you've made? colcon build outputs the following error when attempting to buils occupnacy-fuser

<img src="https://github.com/Monash-Connected-Autonomous-Vehicle/ESDA/assets/95030427/ec0ef308-a5d5-484d-a582-8610540c7552" width="400"/>

Use of hyphen is said to be the issue as per https://answers.ros.org/question/368144/colcon-build-fails-on-newly-created-package/ but I don't want to trial it given you seem to have gotten it built and I might just be missing something.

Rosdeps have al been installed

**Comment by AbBaSaMo** [comment]
@AnthonyZhOon changed package name to using '_' insread of '-' which fixed the issue. If I can confirm you have no unpushed changed, am gonna push.

![Image](https://github.com/Monash-Connected-Autonomous-Vehicle/ESDA/assets/95030427/2c272464-d0ee-4dd7-ac6c-74c8acf34bcb)

**Comment by AbBaSaMo** [comment]
It seems the issue is that the occupancy grids have not been initialized with correct parameters, so based on the following grid1 message, you have a 51x51 grid with only cell 50 occupied, and a resolution of 0 so each cell is 0x0m iirc

![Image](https://github.com/Monash-Connected-Autonomous-Vehicle/ESDA/assets/95030427/17b115a6-6b8f-48f5-a372-659f7c3c6f94)

**Comment by AbBaSaMo** [comment]
@AnthonyZhOon can confirm that this outputs *a* grid. will push after class plus a meeting, and possible add a random generator for cells

![Image](https://github.com/Monash-Connected-Autonomous-Vehicle/ESDA/assets/95030427/aaf887bf-3255-423a-b996-990310a07c85)

**Comment by AbBaSaMo** [comment]
Another thing to note is that you'll probs need a static tf publisher for base_link. Publishing an arbitrary one, but I'll have to confirm if it doesnt work without it

![Image](https://github.com/Monash-Connected-Autonomous-Vehicle/ESDA/assets/95030427/3607a791-5d75-41c1-8f0c-cdafabb6bfb5)

**Comment by AbBaSaMo** [comment]
@AnthonyZhOon pushed fixes and changes, so the remaining tasks for testing are:
- [ ] fix grid publishers so they aren't occupied at just cell 50, make it random or smth
- [ ] test and verify accuracy of fusion
- [ ] consider that the resolution of the grids might differ
